### ISSUE TITLE
Mixer module sometimes hangs calling pygame.quit() on Virtual Machine/ArchLinux

### ISSUE THREAD

**Issue by illume** [pygame/pygame#329]
Originally reported by: **ocket8888 (Bitbucket: [ocket8888](https://bitbucket.org/ocket8888), GitHub: [ocket8888](https://github.com/ocket8888))**

----------------------------------------

Just like the title says, I've never been able to write a pygame script that will exit properly. Even something as simple as

```
#!python

#!/usr/bin/env python3

import pygame
from sys import exit as sysExit

pygame.init()
screen = pygame.display.set_mode((512,512))
screen.fill((255,255,255))

pygame.display.update()

while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.display.update()
            pygame.display.quit()
            pygame.quit()
            exit(0)
            sysExit(0)
            #None of these work, code under them is unreachable but the process hangs
```
hangs. More information is available [here](http://stackoverflow.com/questions/42572322/python-3-pygame-programs-will-not-quit)

----------------------------------------
- Bitbucket: https://bitbucket.org/pygame/pygame/issue/329

Steps to proceed with this:

- [ ] Try the code in the above with pygame 2.0.0.dev10 running in a linux virtual machine.
- [ ] If it is still broken try the code above replacing `pygame.init()` & `pygame.display.init()`

**Related Docs:** https://www.pygame.org/docs/ref/pygame.html

**Comment by surculus12** [comment]
I am experiencing this issue in a 64-bit virtualbox instance (`Version 5.2.12 r122591 (Qt5.6.2)`) and on a Windows 10 Enterprise host (`10.0.15063 Build 15063`).
```
$ uname -a
Linux surculus12-VirtualBox 4.13.0-39-generic #44~16.04.1-Ubuntu SMP Thu Apr 5 16:43:10 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.4 LTS
Release:        16.04
Codename:       xenial
```

Example of program run (`pygame 1.9.3`, `python 3.6.5`):
```
def foo():
    pygame.init()

    # Create window to facilitate QUIT event
    screen = pygame.display.set_mode((640, 480,))

    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()  # Hangs here
                sys.exit()
```

The program hangs inside of `pygame.quit()`. 

Bypassing most pygame initialization via `pygame.display.init()` causes graceful exit.

**Comment by jackjt8** [comment]
I'm experiencing this issue, or at least a similar one, trying to exit Pygame on a Raspberry Pi Zero running Raspbrain Stretch (default python installs). 

`pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.14.34+ #1110 Mon Apr 16 14:51:42 BST 2018 armv6l GNU/Linux
pi@raspberrypi:~ $ python -V
`

`Python 2.7.13
Python 3.5.3 
Pygame 1.9.3`

With Python2 `pygame.quit()` causes it to hang. For whatever reason, SSH'ing and using `sudo kill id`, `sudo kill -9 id`, and `sudo killall python` are ineffective. Likewise, no shutdown method will work if the script is still running. I can only pull the power which may corrupt my SD card. Running script from SSH and then using crtl+c via SSH will "lock up" that SSH window and I need to create a new SSH session to do anything. Switching `pygame.init()` to `pygame.display.init()` does solve the issue for me.

Running the program under Python3 produces a `Segmentation fault` and a black framebuffer. Rerunning the script it will then draw a white box with a movable mouse pointer but it will not auto exit, using crtl+c it will only remove the top half of the box and won't clear the framebuffer. Running script from SSH and using crtl+c via SSH will break the framebuffer as above, but the SSH window will still be usable. Switching `pygame.init()` to `pygame.display.init()` does not solve the issue with Python3 and it acts the same as before.

tl;dr Python2 pygam.exit() is fixed by switching `pygame.init()` to `pygame.display.init()`. Python3 pygame throws a `Segmentation fault` on first run, produces a white box with mouse pointer on all other runs and doesn't auto exit.

**Comment by Bottersnike** [comment]
I have the same issue on Arch Linux x64. My current solution is either to just init the things I need and then `.quit` them and only them or use `os._exit(0)`.

**Comment by nudelbrot** [comment]
Can confirm the issue, also on Arch Linux x64; not a VM.
@jackjt8's solution by switching from `pygame.init()` to `pygame.display.init()` solves the problem, also when using Python3.

**Comment by paunis** [comment]
I have the same issue but it's intermitent. Sometimes my app closes properly, sometimes it hangs. I'm running it in Python 3 on a Linux virtual machine.

**Comment by illume** [comment]
@paunis I wonder which version of pygame you are using? Have you tried `pygame==2.0.0.dev3`?

**Comment by paunis** [comment]
> > 
> @paunis I wonder which version of pygame you are using? Have you tried `pygame==2.0.0.dev3`?

Latest 1.9.6:
Name: pygame
Version: 1.9.6
Summary: Python Game Development
Home-page: https://www.pygame.org
Author: A community project.
Author-email: pygame@pygame.org
License: LGPL

**Comment by MyreMylar** [comment]
So to summarise this issue. This code:

    import sys
    import pygame
    
    pygame.init()
    
    screen = pygame.display.set_mode((640, 480,))
    
    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                sys.exit()

Causes pygame to hang on some system setups. 

I've just tried it on Windows in 1.9.6 and 2.0.0.dev8 and it closes fine in both. There were also one report of the issue being intermittent so I also made this program which will init() and quit() pygame 25 times over 50 seconds:

    import sys
    import pygame
    
    for _ in range(25):
        pygame.init()
    
        screen = pygame.display.set_mode((640, 480))
    
        clock = pygame.time.Clock()
        clock.tick()
        timer = 0.0
    
        running = True
        while running:
            timer += clock.tick()
            if timer > 500:
                pygame.event.post(pygame.event.Event(pygame.QUIT))
                timer = 0.0
    
            pygame.display.update()
    
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    running = False
    
    sys.exit()


Which also works fine for me on Windows in pygame 1.9.6, and pygame 2.0.0.dev8.

Since the majority of the reports were from Virtual Machines of some kind I think that is probably the main area of fruitful research. It also appears to be strongly related to the mixer module, given several reports of people fixing it by not calling `pygame.init()` but just calling `pygame.display.init()`.

As such I'll label it with that information and rename it.

**Comment by MyreMylar** [comment]
Steps to proceed with this:

- [ ] Try the code in the post above with pygame 2.0.0.dev8 running in a virtual machine.
- [ ] If it is still broken try the code above replacing `pygame.init()` & `pygame.display.init()`

**Comment by bmillwood** [comment]
I've seen this already exhibited by just:

```
#!/usr/bin/env python3
import pygame

pygame.mixer.init()
print('quitting now')
pygame.mixer.quit()
print('finished quitting')
```

(immediately printing "quitting now" but never "finished quitting"). FWIW ctrl-C and SIGTERM and ineffective but ctrl-\ (SIGQUIT) does the trick.

For me, it hangs when I'm running `fluidsynth` (a software synthesizer I was running to provide a midi device to do midi stuff with), and interestingly, the hang ends (the program exits normally, printing "finished quitting") when I stop `fluidsynth` (I'm just running `fluidsynth --server --audio-driver=alsa -o audio.alsa.device=hw:0 soundfont.sf2` in the foreground and ctrl-C-ing it when I'm done).

Can strace if you give me some particular syscalls to trace -- the unfiltered output is too much for me to review.

I'm on 1.9.6.

**Comment by MyreMylar** [comment]
Hmm. Do you mean that you are running fluidsynth separately independent from whatever you are doing with pygame - or that you are using fluidsynth specifically in order to improve pygame audio?

I believe that SDL (which pygame is built on) also uses/can use fluidsynth to play audio instead of Timidity. This post from way back in 2011 suggest it will be used by SDL first if it is available before falling back on timidity:

https://discourse.libsdl.org/t/patch-fluidsynth-backend-for-sdl-mixer/18070/15

Though that is from 2011/2012 so the situation may have evolved since. Perhaps some code somewhere along the line is waiting for fluidsynth or Timidity to report that it has shut down and it isn't getting shut down (perhaps because SDL didn't start it, whichever one is being waited for) and so it waits forever.

It's possible that this may be magically resolved by the slow migration of changes from SDL Mixer X back into to SDL Mixer? Though that is just wild speculation. I feel like SDL mixer has not traditionally been frequently updated.


(SDL Mixer X: https://github.com/WohlSoft/SDL-Mixer-X)

Though it could be something else entirely. This might be related:

https://github.com/FluidSynth/fluidsynth/pull/478

Though it post dates the first problems we had reported in here.



You can see the pygame C code called when mixer is quit here:

https://github.com/pygame/pygame/blob/master/src_c/mixer.c#L298-L341

No obvious loops that could be hanging there. I guess the only way to take it further code wise would be building an SDL C application, on a linux system that has the issue, one that essentially does the same thing pygame is doing. Then see if it still hangs on running the quit function, and then from there tracking it down to a specific call to SDL.

Looks like it would be a choice of:

Mix_HaltMusic()
Mix_FreeMusic()
Mix_CloseAudio()
SDL_QuitSubSystem(SDL_INIT_AUDIO)

Searching Github some on thos function calls, I found this thread:
https://github.com/wesnoth/wesnoth/issues/3588

Which sounds very similar (SDL hanging while another audio using program is running) and is highly suggestive that the problem is likely to lie with Mix_CloseAudio().

**Comment by bmillwood** [comment]
I'm playing pygame midi audio through fluidsynth (without it, I don't hear anything at all). I believe I don't have timidity installed.

Indeed it does seem like this C file reproduces my problem:

```c
#include "SDL.h"
#include "SDL_mixer.h"

int main(int argc, char *argv[])
{
  int ret;
  ret = SDL_Init(SDL_INIT_AUDIO);
  if(ret < 0) {
    return 1;
  }
  
  ret = Mix_OpenAudio(MIX_DEFAULT_FREQUENCY, MIX_DEFAULT_FORMAT, 2, 1024);
  if(ret < 0) {
    return 2;
  }

  printf("before Mix_CloseAudio\n");
  Mix_CloseAudio();
  printf("after Mix_CloseAudio\n");
  SDL_Quit();
  return 0;
}

```

(i.e. it hangs on "before Mix_CloseAudio" until I stop fluidsynth).

**Comment by rodetroner** [comment]
I was able to reproduce the bug with
```
import pygame

pygame.mixer.init()
pygame.display.init()
pygame.display.quit()
pygame.mixer.quit()
```
However, with the code below, the program terminates correctly.
```
import pygame

pygame.mixer.init()
pygame.mixer.quit()
```

**Comment by ankith26** [comment]
Reviving old issue, I think we can safely say this is a confirmed bug in `Mix_CloseAudio`, so the next step should be reporting it on `SDL_mixer` github repo with the C code reproducer?

**Comment by MyreMylar** [comment]
I believe it is only with the fluidsynth backend as well, I guess somebody who actually gets the bug should test it (i.e. someone with a linux system and a fluidsynth install)

**Comment by bmillwood** [comment]
I posted about my reproduction case in libsdl-org/SDL_mixer#367 but it's not 100% clear if the other reproductions involve fluidsynth / are affected by the same alsa device thing that I noticed while filing that issue. Would be interested to see other people still affected by this comment on that?

**Comment by mezdahun** [comment]
I use pygame to carry out agent-based simulations for which I run a pygame based simulation instance for a given time, save some generated data, then quit, change some parameters in my framework and restart the process. I can reproduce this issue (or closely related) on Ubuntu 20 in an Oracle VM,  in a non-VM installation and in a Docker container where pygame runs in headless mode. The quit method hangs **when I switch user** (when using docker switching user on the host). This might be relevant to fix the issue so I leave the comment here in case it helps fixing the issue or reproduce it. As described before switching to `pygame.display.init()` from `pygame.init()` seemingly solved the problem and I can keep running my simulations with other users using the same machine.

**Comment by ArtBIT** [comment]
I can reproduce this issue on Ubuntu 22.04.1, SDL 2.30.6, Python 3.10.12 with the following code:
```
import pygame

pygame.mixer.init()
pygame.display.init()
pygame.display.quit()
pygame.mixer.quit()

```

It is intermittent though. Sometimes it quits correctly, sometimes it doesn't.
I've noticed that if my game has sound, it quits correctly, and if it starts silent, it will hang on quit.
In both cases, `pygame.init` returns:
```
pygame.init: 5 pass, 0 fails
```

**Comment by afiskon** [comment]
Here are my two cents.

This issue reproduces on Ubuntu 24.04 LTS, Intel x64.

```
pygame 2.6.1 (SDL 2.28.4, Python 3.12.3)
```
Calling init() only for required sub-packages, e.g. display and font, but not mixer - "solves" the issuse. Alternatively the script can be terminated with `os._exit(0)` as suggested above.

This issue seems to be Linux-specific since everything works fine on MacOS (also Intel x64).

If there is anything else I can help with please let me know. I'm not 100% sure how to debug this further.
